src/Castle.Core/DynamicProxy/Generators/GeneratorUtil.cs

```diff
@@ -35,12 +37,62 @@ public static class GeneratorUtil
 
 			for (var i = 0; i < parameters.Length; i++)
 			{
-				if (!parameters[i].ParameterType.GetTypeInfo().IsByRef)
+				if (IsByRef(parameters[i]) && !IsReadOnly(parameters[i]))
 				{
-					continue;
+					emitter.CodeBuilder.AddStatement(AssignArgument(dereferencedArguments, i, arguments));
 				}
+			}
+
+			bool IsByRef(ParameterInfo parameter)
+			{
+				return parameter.ParameterType.GetTypeInfo().IsByRef;
+			}
 
-				emitter.CodeBuilder.AddStatement(AssignArgument(dereferencedArguments, i, arguments));
+			bool IsReadOnly(ParameterInfo parameter)
+			{
+				// C# `in` parameters are also by-ref, but meant to be read-only.
+				// The section "Metadata representation of in parameters" on the following page
+				// defines how such parameters are marked:
+				//
+				// https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.2/readonly-ref.md
+				//
+				// This poses three problems for detecting them:
+				//
+				//  * The C# Roslyn compiler marks `in` parameters with an `[in]` IL modifier,
+				//    but this isn't specified, nor is it used uniquely for `in` params.
+				//
+				//  * `System.Runtime.CompilerServices.IsReadOnlyAttribute` is not defined on all
+				//    .NET platforms, so the compiler sometimes recreates that type in the same
+				//    assembly that contains the method having an `in` parameter. In other words,
+				//    it's an attribute one must check for by name (which is slow, as it implies
+				//    use of a `GetCustomAttributes` enumeration instead of a faster `IsDefined`).
+				//
+				//  * A required custom modifier `System.Runtime.InteropServices.InAttribute`
+				//    is always present in those cases relevant for DynamicProxy (proxyable methods),
+				//    but not all targeted platforms support reading custom modifiers. Also,
+				//    support for cmods is generally flaky (at this time of writing, mid-2018).
+				//
+				// The above points inform the following detection logic: First, we rely on an IL
+				// `[in]` modifier being present. This is a "fast guard" against non-`in` parameters:
+				if ((parameter.Attributes & (ParameterAttributes.In | ParameterAttributes.Out)) == ParameterAttributes.In)
+				{
+					// Here we perform the actual check. We don't rely on cmods because support
+					// for them is at current too unreliable in general, and because we wouldn't
+					// be saving much time anyway.
+					if (parameter.GetCustomAttributes(false).Any(IsIsReadOnlyAttribute))
+					{
+						return true;
+					}
+				}
+
+				return false;
+
+				bool IsIsReadOnlyAttribute(object attribute)
```

Comment:
😆, I almost thought that might be your reaction!

The idea here is to make the scope of helper functions as local as possible, so that it's clear where they are actually needed. Contrast this with a "global" helper function like `ArgumentsUtil.IsAnyByRef` which, as it turned out, was only used in one single location&mdash;but you couldn't tell by taking a casual glance at its source.

Options:

1. Leave it as it is now (nested local functions).
2. Bring `IsIsReadOnlyAttribute` to the same level as the other local functions to reduce nesting.
3. Turn the three local functions into `private static` methods in the same class (`GeneratorUtil`).
4. Revert to how things were before (i.e. don't have local functions at all).

What's your preference?

Reply:
@jonorossi: 😆, I almost thought that might be your reaction!

The idea here is to make the scope of helper functions as local as possible, so that it's clear where they are actually needed. Contrast this with a "global" helper function like `ArgumentsUtil.IsAnyByRef` which, as it turned out, was only used in one single location&mdash;but you couldn't tell by taking a casual glance at its source.

Options:

1. Leave it as it is now (nested local functions).
2. Bring `IsIsReadOnlyAttribute` to the same level as the two other local functions to reduce nesting.
3. Turn the three local functions into `private static` methods in the same class (`GeneratorUtil`).
4. Revert to how things were before (i.e. don't have local functions at all).

What's your preference?

Reply:
Let's just leave it as is, I didn't really have a preference other than being a little surprised. I think if `CopyOutAndRefParameters` was non-void it would look strange because you'd have to try to match up returns.